pkg/trace/agent/normalizer.go

```diff
@@ -41,13 +42,31 @@ var (
 	Year2000NanosecTS = time.Date(2000, time.January, 1, 0, 0, 0, 0, time.UTC).UnixNano()
 )
 
+// fallbackServiceNames holds a map of default service names to use. It maps
+// tracer language to service name.
+var fallbackServiceNames sync.Map
+
+// fallbackService returns the fallback service name for a service
+// belonging to language lang.
+func fallbackService(lang string) string {
+	if lang == "" {
+		return DefaultServiceName
+	}
+	if v, ok := fallbackServiceNames.Load(lang); ok {
```

Comment:
Alternatively, we could use a structure like `normalizer`, 1 per goroutine, and then add a map `fallbackServiceNames` in this structure.

That way, no need to protect it with `sync` operations.

Reply:
i would try hardcoding our common languages and only fallback to a map on a language miss.

if lang  == "java" { return "unnamed-lang-service"}` this should make the code almost invisible for the vast majority of spans and skip the overhead of sync.map which still does locking under the hood. 

```
switch lang {
case "java": return "dd-trace-java"
case "go": return "dd-trace-go"
default:
// map code
}

```

Reply:
I dropped this suggestion because once the langs get added to the map its virtually free to retrieve them. It's an `atomic.LoadUintPtr` call.